README.md
 # forxpy

Package to access daily exchange rates and forex conversion created by Group 6: Dhruvi Nishar, Mohammad Reza Nabizadeh, Hongjian Li, and Stepan Zaiatc.

`forxpy` allows users to easily convert currency rates by utilizing data from [Bank of Canada](https://www.bankofcanada.ca/rates/exchange/daily-exchange-rates/). The package supports multiple currencies and provides historical exchange rates. It also has the ability to make currency conversions with the use of a simple function call, making it easy for developers to integrate into their projects. Additionally, it provides a way to get historical exchange rates in a chart, which makes it useful for dashboard building. Overall, **forxpy** is a powerful and flexible package for handling currency conversions in Python.

## Installation

```bash
$ pip install forxpy
```

## Usage

- TODO

### Functions included in the package:

- **retrieve_data()**: Retrieve historical daily currency exchange rate data for Canadian Dollars in CSV format from the Bank of Canada website.
- **currency_convert**: The conversion rate is based on the average exchange rate by the 4:00 pm ET of the closest business day.
- **fastest_slowest_currency()**: This function takes start and end date as input and returns a  list of two lists containing the fastest and slowest growing currency exchange rate in relation to Canadian Dollar along with their current exchange rate in the provided date range.
- **plot_historical()**:Plots the historical rate of the entered currencies within a specific period of time.

**forxpy** can be a useful tool for many industries that require currency conversions. Here are a few examples:
1. **Finance and Banking**: Banks and financial institutions often need to convert currencies for international transactions, and "forxpy" can provide accurate and up-to-date exchange rates for these purposes.
2. **E-commerce**: Online retailers that sell internationally may need to display prices in multiple currencies, and "forxpy" can help them convert the prices in real-time.
3. **Travel and Tourism**: Travel agencies and booking websites may need to convert currency for pricing and budgeting purposes, and "forxpy" can provide them with accurate exchange rates.
4. **Data Analysis**: Companies that conduct global business can use "forxpy" to collect and analyze historical exchange rate data, which can be useful for making business decisions and forecasting.
5. **Cryptocurrency**: As the cryptocurrency markets are decentralized and global, the package can be used to convert cryptocurrency prices to different fiat currencies, which is helpful for investors and traders.

Overall, "forxpy" can be a valuable tool for any industry that needs to handle currency conversions in Python.

## Similar Packages

1. [`forex-python`](https://pypi.org/project/forex-python/): Used to retrieve foreign currency exchange rates and currency conversion.
2. [`CurrencyConverter`](https://pypi.org/project/CurrencyConverter/): This is a currency converter package that uses historical rates against a reference currency (Euro)


## Contributing
**Authors**: Dhruvi Nishar, Mohammad Reza Nabizadeh, Hongjian Li, and Stepan Zaiatc.

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`forxpy` was created by Group 6 - Dhruvi Nishar, Mohammad Reza Nabizadeh, Hongjian Li, and Stepan Zaiatc. It is licensed under the terms of the MIT license.

## Credits

`forxpy` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v2.0.0 (2023-01-29)
### Feature
* **interactive_plotting:** Add the interactive plotting                 to forxpy module ([`416c23b`](https://github.com/UBC-MDS/forxpy/commit/416c23be9bfca605fd85da3e05f024257d949ce5))

### Breaking
* plotting module will be updated to the interactive module. ([`416c23b`](https://github.com/UBC-MDS/forxpy/commit/416c23be9bfca605fd85da3e05f024257d949ce5))

## v1.0.0 (2023-01-24)
### Feature
* **mod_plotting:** Move code from plotting module                  to pycounts module ([`21579ea`](https://github.com/UBC-MDS/forxpy/commit/21579ea3692d857a10343bb30dd619ff83fb32e1))

### Breaking
* plotting module hass be added to this release. ([`21579ea`](https://github.com/UBC-MDS/forxpy/commit/21579ea3692d857a10343bb30dd619ff83fb32e1))

## v0.0.4 (2023-01-24)


## v0.1.0 (11/01/2023)

- First release of `forxpy`!


tests/test_forxpy.py
import pytest
from forxpy.forxpy import *
from datetime import datetime

def test_retrieve_data():

    data = retrieve_data(export_csv = False)

    """ Test whether the first row has been dropped. """
    assert data.index[0] == 1, "First row is not dropped"
    
    """ Test whether the required columns have been dropped. """
    assert 'FXMYRCAD' not in data.columns, "FXMYRCAD column is not dropped"
    assert 'FXTHBCAD' not in data.columns, "FXTHBCAD column is not dropped"
    assert 'FXVNDCAD' not in data.columns, "FXVNDCAD column is not dropped"
    
    """ Test whether the output is pandas DataFrame data type. """
    if not isinstance(data, pd.DataFrame):
        raise TypeError("Output is not pd.DataFrame data type")


def test_plot_historical():
    sbs_plot = plot_historical('2020-05-23', '2022-05-30', 'USD', 'CAD')
    plt_objs = sbs_plot.to_dict()

    """ Test whether the final plot has been formed """
    plt_title = 'How many CAD does 1 USD worth?'
    assert sbs_plot.title == plt_title, 'The final plot has not formed correctly'
    
    """ Test whether the subplot has been formed correctly """
    assert plt_objs['vconcat'][0]['layer'][0]['mark'] == 'line', 'Chart type is not line'
    assert plt_objs['vconcat'][0]['layer'][0]['encoding']['x']['field'] == 'date', 'The X axis is not date'
    assert plt_objs['vconcat'][0]['layer'][0]['encoding']['y']['field'] == 'USD/CAD', 'Rates are not numeric'

def test_fastest_slowest_currency():
    start_date = '2019-05-23'
    end_date = '2022-05-30'

    res = fastest_slowest_currency(start_date, end_date)

    """ Test whether the start and end date are correct """
    start = datetime.strptime(start_date, '%Y-%m-%d')
    end = datetime.strptime(end_date, '%Y-%m-%d')
    assert (end - start).days > 0, 'The date range is inaccurate'

    """ Test if it returns a list type """
    assert type(res) == list, 'The function does not return a list of lists accurately'

    """ Test 2: Check if the function is returning the correct fastest currency """
    assert fastest_slowest_currency('2019-05-23', '2022-05-30')[0][0] == 'TRY'

    """ Test 3: Check if the function is returning the correct slowest currency """
    assert fastest_slowest_currency('2019-05-23', '2022-05-30')[1][0] == 'IDR'

def test_currency_convert():
    df_conv_rates = retrieve_data()
    names = {'AUD':1,'BRL':2,'CNY':3, 'EUR':4, 'HKD':5, 'INR':6, 
             'IDR':7, 'JPY':8, 'MXN':9, 'NZD':10, 'NOK':11, 'PEN':12, 
             'RUB':13, 'SAR':14, 'SGD':15, 'ZAR':16, 'KRW':17, 'SEK':18, 
             'CHF':19, 'TWD':20, 'TRY':21, 'GBP':22, 'USD':23} #initializations
    
    assert type(currency_convert(2, 'CAD', 'USD')) is float or type(currency_convert(1, 'CAD', 'CAD')) is int , " The output data type is wrong!" # Test for the output data type
    
    assert currency_convert(2, 'CAD','CAD') == 2,"Wrong output value" #Test for exchanging the same currency
    
    assert currency_convert(2, 'CNY','CAD') == round(2*float(df_conv_rates.iloc[-1][names['CNY']]),3),"Wrong output value!" #Test for converting another currency to CAD
    
    assert currency_convert(2, 'CAD', 'CNY') == round(2/(float(df_conv_rates.iloc[-1][names['CNY']])),3), "Wrong output value!" #Test for converting CAD to another currency

src/forxpy/forxpy.py
import os
import pandas as pd
import altair as alt
from datetime import datetime
import re

def retrieve_data(export_csv = False):

    """
    Retrieve historical daily currency exchange rates data for Canadian Dollar 
    in CSV format from Bank of Canada website. 
    The function pre-processes and cleans the data to transform it into a more usable format.     
    
    Parameters
    ----------
    export_csv : bool
        If the value is False then only display the data frame, If the value is True then write the file to the current working directory. 

    Returns
    -------
    file :
        cleaned and processed dataframe and csv file that includes historical 
        data on currency exchange rates
        
    >>> retrieve_data(export_csv = False)

    """
    # Read CSV file and reset the index
    url = 'https://raw.githubusercontent.com/mrnabiz/forx_source/main/data/raw/raw_data_cad.csv'
    data_raw = (pd.read_csv(url, delimiter=",")[38:]).reset_index()
    
    # Setting the first row as column names
    data_raw.columns = data_raw.iloc[0]
    data = data_raw.iloc[:, 1:]
    
    # Drop the first row of data
    data = data.drop(data.index[0])
    
    # Drop "FXMYRCAD", "FXTHBCAD", "FXVNDCAD" columns with many NA values
    data = data.drop(labels=["FXMYRCAD", "FXTHBCAD", "FXVNDCAD"], axis=1)
    
    
    # Convert date column to datetime format
    data['date'] = pd.to_datetime(data['date'])
    
    # Creating list of column names
    col_list = data.columns.tolist()
    
    # Remove the date column from the list
    col_list.remove('date')
    
    # Convert all columns in the list to numeric data type
    data[col_list] = data[col_list].apply(pd.to_numeric)
    
    # Wrangling column labels
    data.columns = data.columns.str.replace("FX", "")
    data.columns = data.columns.str.replace("CAD", "")
    
    # Add "CAD" column and assign a value of 1.0 to eahc row
    data['CAD'] = 1.0
    

    if export_csv == True:
        # Saving dataframe as CSV file if output=True
        data.to_csv("data_raw.csv")
    else:
        # Only display dataframe if output=False
        return data
    
    return data


def fastest_slowest_currency(start_date, end_date):
    """
    This function takes currency exchange rates data as input and returns a 
    list of two strings containing the fastest and slowest growing currency 
    exchange rate in relation to Canadian Dollar.
    The data provided contains currency code in the format FX***CAD, 
    the average exchange rate and the date.
    
    Parameters
    ----------
    start_date : string '%YYYY-%mm-%dd'
	    inputted starting date in the format specified '%YYYY-%mm-%dd'
	end_date : string '%YYYY-%mm-%dd'
	    inputted ending date in the format specified '%YYYY-%mm-%dd'

    Returns
    -------
    list
        list of lists containing the fastest currency name and its current 
        exchange rate with CAD and the slowest currency with its current exchange 
        rate with CAD for the specified date range.
    
    Examples
    >>> fastest_slowest_currency('2019-05-23', '2022-05-30')
    [['EUR', 1.4545], ['IDR', 8.9e-05]]
    """ 
    import warnings
    warnings.simplefilter(action='ignore', category=FutureWarning)

    # Check for invalid date format
    if not re.match(r"^\d{4}-\d{2}-\d{2}$", start_date) or not re.match(r"^\d{4}-\d{2}-\d{2}$", end_date):
        raise ValueError("Invalid date format. Please enter dates in the format '%YYYY-%mm-%dd'.")
    
    # Check for invalid date range
    start = datetime.strptime(start_date, '%Y-%m-%d')
    end = datetime.strptime(end_date, '%Y-%m-%d')
    if start > end:
        raise ValueError("Invalid date range. Please ensure that the start date is before the end date.")

    # Extracting the data split to calculate the fastest and slowest currency for the given range
    data = retrieve_data()
    df = data[(data['date'] >= start) & (data['date'] <= end)]

    # Check for empty data
    if data.empty:
        raise ValueError("No data available for the specified date range.")

    # Computing the fastest growing currency and the slowest growing currency for the given range
    tepm = df[:1] 
    tepm = tepm.append(df[-1:], ignore_index=True)
    diff = tepm.diff()[-1:]
    diff = diff.abs()
    
    nums = pd.to_numeric(diff.drop(columns=['date', 'CAD']).loc[1])
    fastestcurr = nums.idxmax()
    slowestcurr = nums.idxmin()

    # calculates and stores the increase and decrease overall
    fastdiff = nums.max()
    slowdiff = nums.min()

    # Extracting the current rate of the slowest and the fastest currencies
    slow_current_rate = data.loc[data.shape[0]][slowestcurr]
    fast_current_rate = data.loc[data.shape[0]][fastestcurr]

    # returning the computed values
    return [[fastestcurr, fast_current_rate], [slowestcurr, slow_current_rate]]

def currency_convert(value, currency1, currency2):
    """
    This function takes a currency value and the currency type 
    to be converted to as input and returns the converted currency 
    value as per the current conversion rate.
    
    Parameters
    ----------
    value: float
        The value of the original currency to be converted

    currency1: str
        The type of currency originally
    
    currency2: str
        The type of currency that the currency1 will be converted to
    
    Returns
    -------
    converted: numeric
        Returns converted numeric currency
    
    Examples
    >>> currency_convert(23, 'USD', 'CAD')
    """
    df_conv_rates = retrieve_data('https://raw.githubusercontent.com/mrnabiz/forx_source/main/data/raw/raw_data_cad.csv')
    
    names = {'AUD':1,'BRL':2,'CNY':3, 'EUR':4, 'HKD':5, 'INR':6, 'IDR':7, 'JPY':8, 'MXN':9, 'NZD':10, 'NOK':11, 'PEN':12, 
         'RUB':13, 'SAR':14, 'SGD':15, 'ZAR':16, 'KRW':17, 'SEK':18, 'CHF':19, 'TWD':20, 'TRY':21, 'GBP':22, 'USD':23} # helps to locate the data in the df
    
    if currency1 not in names: # check whether the currency to be converted is in our data
        if currency1 != 'CAD':
            raise ValueError("The currency to be converted is invalid!")
        
    if currency2 not in names: # check whether the currency to be converted to is in our data
        if currency2 != 'CAD':
            raise ValueError("The currency to be converted to is invalid!")
    if value <= 0:
        raise ValueError("Please enter an positive amount!")
    
    if currency2 == 'CAD' : # situation of convert a specific currency to CAD
        if currency1 == 'CAD':
            return(round(value,3)) 
        
        else:
            amount = round(value*float(df_conv_rates.iloc[-1][names[currency1]]),3)
            return amount 
    
    elif currency1 == 'CAD': # situation of convert CAD to a specific currency
        if currency2 == 'CAD':
            return(round(value,3))
        
        else: 
            amount = round(value/(float(df_conv_rates.iloc[-1][names[currency2]])),3)
            return amount
    
    else: 
            amount = round(value*(float(df_conv_rates.iloc[-1][names[currency1]])) # situation of convert a specific currency to another currency 
                 /(float(df_conv_rates.iloc[-1][names[currency2]])),3)              # which are neither CAD 
            
            return amount
    

def plot_historical(start_date, end_date, currency1, currency2):
    """
    Plots the historical rate of the entered currencies within a specific period
    of time.

    Parameters
    ----------
	start_date : string '%YYYY-%mm-%dd'
	    inputted starting date in the format specified '%YYYY-%mm-%dd'
	end_date : string '%YYYY-%mm-%dd'
	    inputted ending date in the format specified '%YYYY-%mm-%dd'
    currency1 : str
        The type of based currency asked for plotting
    currency2 : str
        The type of exchange currency asked for plotting

    Returns
    -------
    plot object
        A plot showing the performance of the currency.

    Examples
    --------
    >>> plot_historical('2020-05-23', '2022-05-30', 'USD', 'CAD')
    """
    import warnings
    warnings.simplefilter(action='ignore', category=FutureWarning)

    # Data filtration
    start = datetime.strptime(start_date, '%Y-%m-%d')
    end = datetime.strptime(end_date, '%Y-%m-%d')
    ratio = currency1 + '/' + currency2
    data = retrieve_data()
    data[ratio] = data[currency1]/data[currency2]
    data_plt = data[(data['date'] >= start) &
                    (data['date'] <= end)][['date', ratio]]
    
    # Building the base chart
    base_chart = alt.Chart(data_plt).mark_line().encode(
        alt.X('date', title='Date'),
        alt.Y(ratio, scale=alt.Scale(zero=False)),
        tooltip=['date', ratio]).properties(width=900, height=200)
    dot_line_chart = base_chart + base_chart.mark_point(size=2)

    # Building interactivity
    brush = alt.selection_interval(encodings=['x'])
    lower_chart = base_chart.properties(height=60).add_selection(brush)
    upper_chart = dot_line_chart.encode(alt.X('date:T', 
                                            scale=alt.Scale(domain=brush)))

    plt_title = 'How many ' + currency2 + ' does 1 ' + currency1 + ' worth?'
    sbs_plot = (upper_chart & lower_chart).properties(
        title=plt_title
                                        ).configure_title(
        fontSize=18, font='Cambria', anchor='start').configure_axis(
            labelFontSize=10, titleFontSize=10, 
            labelFont='Cambria', titleFont='Cambria'
            )
    
    # Unit tests to test the function input
    assert end >= data['date'].min(), 'The end date is out of the range'
    assert start <= data['date'].max(), 'The start date is out of the range'
    assert currency1 in data.columns.to_list(), 'Currency 1 is not supported'
    assert currency2 in data.columns.to_list(), 'Currency 1 is not supported'

    # Unit tests to test the plot object
    assert base_chart.to_dict()['mark'] == 'line', 'Chart type is not line'
    assert base_chart.to_dict()['encoding']['x']['type'] == 'temporal', 'Datetype is not temporal'
    assert base_chart.to_dict()['encoding']['y']['type'] == 'quantitative', 'Rates are not numeric'
    assert sbs_plot.title == plt_title, 'The final plot has not formed correctly'

    return sbs_plot